File: src/leetcode/third/Code_043_MultiplyStrings.java
package leetcode.third;

/**
 * 字符串相乘
 *
 * @since 2020-5-28 Thursday 14:30 - 14:50
 */
public class Code_043_MultiplyStrings {
    static String multiply(String num1, String num2) {
        int l1 = num1.length(), l2 = num2.length();
        int m = num1.length() - 1, n = num2.length() - 1;
        String res = "0";
        for (; m >= 0; m--) {
            for (; n >= 0; n--) {
                int tmp = (num1.charAt(m) - '0') * (num2.charAt(n) - '0');
                StringBuilder addNumber = new StringBuilder();
                addNumber.append(tmp);
                for (int i = 0; i < (l1 - 1 - m + l2 - 1 - n); i++) {
                    addNumber.append(0);
                }
                res = add(addNumber.toString(), res);
            }
        }
        return res;
    }

    static private String add(String num1, String num2) {
        int carry = 0;
        StringBuilder sb = new StringBuilder();
        int m = num1.length() - 1, n = num2.length() - 1;
        while (m >= 0 || n >= 0) {
            int x = 0, y = 0;
            if (m >= 0) {
                x = num1.charAt(m) - '0';
                m--;
            }
            if (n >= 0) {
                y = num2.charAt(n) - '0';
                n--;
            }
            int tmp = x + y + carry;
            carry = tmp > 9 ? 1 : 0;
            sb.append(tmp % 10);
        }
        if (carry == 1) sb.append(1);
        return sb.reverse().toString();
    }

    public static void main(String[] args) {
        System.out.println(multiply("1", "0"));
        System.out.println(multiply("999", "0"));
        System.out.println(multiply("12", "21"));
        System.out.println(multiply("22", "55"));
        System.out.println(multiply("2", "456"));
        System.out.println(multiply("123", "456"));
    }
}

File: src/cpp-leetcode/first/Code_043_MultiplyStrings.cpp
#include <algorithm>
#include <iostream>
#include <string>

using std::string;

/**
 * multiply strings
 *
 * @since 2020-9-27 Sunday 19:10 - 20:09
 */
class Solution {
    string add(string n1, string n2) {
        string res = "";
        int carry = 0;
        int i = n1.size() - 1, j = n2.size() - 1;
        while (i >= 0 && j >= 0) {
            int sum = n1[i--] - '0' + n2[j--] - '0' + carry;
            carry = sum >= 10 ? 1 : 0;
            res += std::to_string(sum % 10);
        }
        if (i < 0) {
            while (j >= 0) {
                int sum = n2[j--] - '0' + carry;
                carry = sum >= 10 ? 1 : 0;
                res += std::to_string(sum % 10);
            }
        }
        if (j < 0) {
            while (i >= 0) {
                int sum = n1[i--] - '0' + carry;
                carry = sum >= 10 ? 1 : 0;
                res += std::to_string(sum % 10);
            }
        }
        if (carry == 1) {
            res.append("1");
        }
        std::reverse(res.begin(), res.end());
        std::cout << "res: " << res << std::endl;
        return res;
    }

   public:
    string multiply(string n1, string n2) {
        string res = "";
        if (n1.size() == 0 || n2.size() == 0) {
            return res;
        }
        if (n1 == "0" || n2 == "0") {
            return "0";
        }
        for (int i = n1.size() - 1; i >= 0; i--) {
            string tmp = "";
            int m = n1[i] - '0', carry = 0;
            for (int j = n2.size() - 1; j >= 0; j--) {
                int product = m * (n2[j] - '0') + carry;
                carry = product / 10;
                tmp += std::to_string(product % 10);
            }
            if (carry > 0) {
                tmp += std::to_string(carry);
            }
            std::reverse(tmp.begin(), tmp.end());
            for (int k = 0; k < n1.size() - 1 - i; k++) {
                tmp.append("0");
            }
            std::cout << "tmp: " << tmp << std::endl;
            res = add(res, tmp);
        }
        return res;
    }
};

int main(int argc, char const *argv[]) {
    Solution s;
    string s1 = "123", s2 = "456", s3 = "0";
    std::cout << s.multiply(s1, s2) << std::endl;
    std::cout << s.multiply(s1, s3) << std::endl;
}
